File: Wikipedia/Code/WKWebView+LoadAssetsHtml.m
#import "WKWebView+LoadAssetsHtml.h"
#import "WMFProxyServer.h"
#import "Wikipedia-Swift.h"

static const NSTimeInterval WKWebViewLoadAssetsHTMLRequestTimeout = 60; //60s is the default NSURLRequest timeout interval

@implementation WKWebView (LoadAssetsHtml)

- (void)loadHTMLFromAssetsFile:(NSString *)fileName scrolledToFragment:(NSString *)fragment {
    if (!fileName) {
        DDLogError(@"attempted to load nil file");
        return;
    }

    fragment = fragment ? fragment : @"top";
    NSURL *requestURL = [[WMFProxyServer sharedProxyServer] proxyURLForRelativeFilePath:fileName fragment:fragment];
    NSURLRequest *request = [NSURLRequest requestWithURL:requestURL cachePolicy:NSURLRequestReloadIgnoringCacheData timeoutInterval:WKWebViewLoadAssetsHTMLRequestTimeout];
    [self loadRequest:request];
}

- (void)loadHTML:(NSString *)string baseURL:(NSURL *)baseURL withAssetsFile:(NSString *)fileName scrolledToFragment:(NSString *)fragment padding:(UIEdgeInsets)padding theme:(WMFTheme *)theme {
    if (!string) {
        string = @"";
    }

    WMFProxyServer *proxyServer = [WMFProxyServer sharedProxyServer];

    if (!proxyServer.isRunning) {
        [proxyServer start];
    }

    string = [proxyServer stringByReplacingImageURLsWithProxyURLsInHTMLString:string withBaseURL:baseURL targetImageWidth:self.window.screen.wmf_articleImageWidthForScale];

    NSString *localFilePath = [[self getAssetsPath] stringByAppendingPathComponent:fileName];

    if (!localFilePath) {
        return;
    }

    NSString *fileContents = [NSMutableString stringWithContentsOfFile:localFilePath
                                                              encoding:NSUTF8StringEncoding
                                                                 error:nil];

    NSNumber *fontSize = [[NSUserDefaults wmf_userDefaults] wmf_articleFontSizeMultiplier];

    NSAssert([fileContents componentsSeparatedByString:@"%@"].count == (5 + 1), @"\nHTML template file does not have required number of percent-ampersand occurences (5).\nNumber of percent-ampersands must match number of values passed to 'stringWithFormat:'");

    NSString *stringToInjectIntoHeadTag = [self stringToInjectIntoHeadTagWithFontSize:fontSize baseURL:baseURL theme:theme];

    // index.html and preview.html have 5 "%@" subsitition markers. Replace these with actual content.
    NSString *templateAndContent = [NSString stringWithFormat:fileContents, stringToInjectIntoHeadTag, @(padding.top), @(padding.left), @(padding.right), string];

    NSUInteger hash = [[baseURL wmf_articleDatabaseKey] hash];
    NSString *requestPath = [NSString stringWithFormat:@"%lu-%@", (unsigned long)hash, fileName];
    [proxyServer setResponseData:[templateAndContent dataUsingEncoding:NSUTF8StringEncoding] withContentType:@"text/html; charset=utf-8" forPath:requestPath];

    [self loadHTMLFromAssetsFile:requestPath scrolledToFragment:fragment];
}

- (NSString *)getAssetsPath {
    return [WikipediaAppUtils assetsPath];
}

- (NSString *)stringToInjectIntoHeadTagWithFontSize:(NSNumber *)fontSize baseURL:(NSURL *)baseURL theme:(WMFTheme *)theme {

    // The 'theme' and 'compatibility' calls are deliberately injected specifically into the head tag via an inline script because:
    //      "... inline scripts are fetched and executed immediately, before the browser continues to parse the page"
    //      https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script
    //
    //  This ensures all theme settings are in place before any page rendering occurs.
    //
    // 'compatibility.enableSupport()'
    //      Needs to happen only once but *before* body elements are present and before
    //      calling 'themes.setTheme()'.
    //
    // 'themes.setTheme()'
    //      Needs to happen before body elements are present so these will appear with
    //      correct theme colors already set. (This method is also used to changes themes,
    //      but changing themes doesn't require 'compatibility.enableSupport()' or
    //      'themes.classifyElements()' be called again.)
    //
    // Reminder:
    //      We don't want to use 'addUserScript:' with WKUserScriptInjectionTimeAtDocumentEnd for this because
    //      it happens too late - at 'DocumentEnd'. We want the colors to be set before this so there is never
    //      a flickering color change visible to the user. We can't use WKUserScriptInjectionTimeAtDocumentBegin
    //      because this fires before any of the head tag contents are resolved, including references to our JS
    //      libraries - we'd have to make a larger set of changes to make this work.

    return [NSString stringWithFormat:@""
                                       "\n<style type='text/css'>"
                                       "\n    body {"
                                       "\n        -webkit-text-size-adjust: %@;"
                                       "\n    }"
                                       "\n</style>"
                                       "\n<base href=\"%@\">"
                                       "\n<script type='text/javascript'>"
                                       "\n    window.wmf.compatibility.enableSupport(document);"
                                       "\n    %@"
                                       "\n</script>",
                                      [NSString stringWithFormat:@"%ld%%", (long)fontSize.integerValue], baseURL.absoluteString, [WKWebView wmf_themeApplicationJavascriptWith:theme]];
}

@end

File: Wikipedia/Code/WKWebView+LoadAssetsHtml.swift
import WebKit
import WMF

let WKWebViewLoadAssetsHTMLRequestTimeout: TimeInterval = 60; //60s is the default NSURLRequest timeout interval

extension WKWebView {
    
    // Loads contents of fileName. Assumes the file is in the "assets" folder.
    @objc func loadHTMLFromAssetsFile(_ fileName: String?, scrolledToFragment fragment: String?) {
        guard
            let fileName = fileName,
            let requestURL = WMFProxyServer.shared().proxyURL(forRelativeFilePath: fileName, fragment: fragment ?? "top")
        else {
            DDLogError("attempted to load nil fileName or requestURL");
            return
        }
        self.load(URLRequest.init(url: requestURL, cachePolicy: .reloadIgnoringCacheData, timeoutInterval: WKWebViewLoadAssetsHTMLRequestTimeout))
    }
    
    // Loads html passed to it injected into html from fileName.
    @objc func loadHTML(_ string: String?, baseURL: URL?, withAssetsFile fileName: String?, scrolledToFragment fragment: String?, padding: UIEdgeInsets, theme: Theme) {
        guard
            let fileName = fileName,
            let baseURL = baseURL,
            let articleDatabaseKey = baseURL.wmf_articleDatabaseKey
        else {
            DDLogError("window, proxyServer, baseURL or fileName not found");
            return
        }
        
        let proxyServer = WMFProxyServer.shared()
        if proxyServer.isRunning == false {
            proxyServer.start()
        }

        let localFilePath = (WikipediaAppUtils.assetsPath() as NSString).appendingPathComponent(fileName)
        guard let fileContents = try? String(contentsOfFile: localFilePath, encoding: String.Encoding.utf8) else {
            DDLogError("\(localFilePath) contents not found");
            return
        }

        assert(fileContents.split(separator: "@").count == (6 + 1), """
                HTML template file does not have required number of percent-ampersand occurences (5).
                Number of percent-ampersands must match number of values passed to 'stringWithFormat:'
        """)
        
        let headTagAddition = stringToInjectIntoHeadTag(fontSize: UserDefaults.wmf.wmf_articleFontSizeMultiplier(), baseURL: baseURL, theme: theme)
        
        var siteCSSLink = ""
        if let baseSite = baseURL.wmf_site?.absoluteString {
            siteCSSLink = """
            <link href="\(baseSite)/api/rest_v1/data/css/mobile/site" rel="stylesheet" type="text/css"></link>
            """
        }
        
        // index.html and preview.html have 6 "%@" subsitition markers. Replace these with actual content.
        let templateAndContent = String(format: fileContents, siteCSSLink, headTagAddition, padding.top as NSNumber, padding.left as NSNumber, padding.right as NSNumber, string ?? "")
        
        let requestPath = "\(articleDatabaseKey.hash)-\(fileName)"
        proxyServer.setResponseData(templateAndContent.data(using: String.Encoding.utf8), withContentType: "text/html; charset=utf-8", forPath: requestPath)
        
        loadHTMLFromAssetsFile(requestPath, scrolledToFragment: fragment)
    }
    
    fileprivate func stringToInjectIntoHeadTag(fontSize: NSNumber, baseURL: URL, theme: Theme) -> String {
        /*
        The 'theme' and 'compatibility' calls are deliberately injected specifically into the head tag via an inline script because:
             "... inline scripts are fetched and executed immediately, before the browser continues to parse the page"
             https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script
        
         This ensures all theme settings are in place before any page rendering occurs.
        
        'compatibility.enableSupport()'
             Needs to happen only once but *before* body elements are present and before
             calling 'themes.setTheme()'.
        
        'themes.setTheme()'
             Needs to happen before body elements are present so these will appear with
             correct theme colors already set. (This method is also used to changes themes,
             but changing themes doesn't require 'compatibility.enableSupport()' or
             'themes.classifyElements()' be called again.)
        
        Reminder:
             We don't want to use 'addUserScript:' with WKUserScriptInjectionTimeAtDocumentEnd for this because
             it happens too late - at 'DocumentEnd'. We want the colors to be set before this so there is never
             a flickering color change visible to the user. We can't use WKUserScriptInjectionTimeAtDocumentBegin
             because this fires before any of the head tag contents are resolved, including references to our JS
             libraries - we'd have to make a larger set of changes to make this work.
        */
        
        return """
            <style type='text/css'>
                body {
                    -webkit-text-size-adjust: \(fontSize)%;
                }
            </style>
            <base href="\(baseURL.absoluteString)">
            <script type='text/javascript'>
                window.wmf.compatibility.enableSupport(document)
                window.wmf.platform.classify(window)
                \(WKWebView.wmf_themeApplicationJavascript(with: theme))
            </script>
            """
    }
}
